<a href="https://colab.research.google.com/github/rabulhasan/CIS-3920/blob/main/associations_rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd 
import numpy as np
transactions = 10000 
itemsets = [ ['battery', 6000], ['sunscreen', 5000], ['sandals', 4000],
            ['bowls', 2000],['battery, sunscreen', 1500], ['battery, sandals', 1000],
            ['battery, bowls', 250], ['battery, sunscreen, sandals',600], ] 
data = np.array(itemsets)
# database 
cols = ('itemset', 'appear_count') 
count = pd.DataFrame(data, columns=cols) 
count

itemset appear_count
0                      battery         6000
1                    sunscreen         5000
2                      sandals         4000
3                        bowls         2000
4           battery, sunscreen         1500
5             battery, sandals         1000
6               battery, bowls          250
7  battery, sunscreen, sandals          600

In [3]:
# check support value of each itemset 
support = lambda x, num: float(x) / num 
count['support'] = [support(n, transactions) for n in count.appear_count] 
count

itemset appear_count  support
0                      battery         6000    0.600
1                    sunscreen         5000    0.500
2                      sandals         4000    0.400
3                        bowls         2000    0.200
4           battery, sunscreen         1500    0.150
5             battery, sandals         1000    0.100
6               battery, bowls          250    0.025
7  battery, sunscreen, sandals          600    0.060

In [4]:
# check whether an itemset is frequent or not
min_support = 0.05
frequent = lambda x: 'yes' if support(x,
transactions) >= min_support else 'no'
count['frequent?'] = [frequent(n) for n in count.appear_count]


In [5]:
# show result
print('minimum support: {}'.format(min_support)) 
count

minimum support: 0.05


itemset appear_count  support frequent?
0                      battery         6000    0.600       yes
1                    sunscreen         5000    0.500       yes
2                      sandals         4000    0.400       yes
3                        bowls         2000    0.200       yes
4           battery, sunscreen         1500    0.150       yes
5             battery, sandals         1000    0.100       yes
6               battery, bowls          250    0.025        no
7  battery, sunscreen, sandals          600    0.060       yes

C) Confidence value
Rule:
Confidence(X→Y)=support(X∧Y)support(X)Confidence(X→Y)=support(X∧Y)support(X)
So:
Confidence({battery}
→{sunscreen})=Support(battery∧sunscreen)Support(battery)=0.150.6=0.25Confidence({battery}
→{sunscreen})=Support(battery∧sunscreen)Support(battery)=0.150.6=0.25
Confidence({battery,sunscreen}→{sandals})=Support({battery,sunscreen}
∧sandals)Support({battery,sunscreen})=0.060.15=0.4Confidence({battery,sunscreen}
→{sandals})=Support({battery,sunscreen}∧sandals)Support({battery,sunscreen})=0.060.15=0.4
The rule {battery,sunscreen}→{sandals}{battery,sunscreen}→{sandals} is more interesting with 40%
confidence.

In [6]:
# get support value from dataframe 
get_support = lambda item: count.support[count.itemset ==item].iloc[0] 
# calculate confidence 
confidence = lambda support_X, support_XY: support_XY / support_X
battery = get_support('battery') 
battery_sunscreen = get_support('battery, sunscreen')
battery_sunscreen_sandals = get_support('battery, sunscreen, sandals')
print(confidence(battery, battery_sunscreen))
print(confidence(battery_sunscreen, battery_sunscreen_sandals))

0.25
0.4


In [7]:
del count['frequent?']
count


itemset appear_count  support
0                      battery         6000    0.600
1                    sunscreen         5000    0.500
2                      sandals         4000    0.400
3                        bowls         2000    0.200
4           battery, sunscreen         1500    0.150
5             battery, sandals         1000    0.100
6               battery, bowls          250    0.025
7  battery, sunscreen, sandals          600    0.060

C*k*: Set of candidates k-itemset.
L*k*: Set of k-itemset that satisfies the minimum support "frequent itemsets".

In our case, the minimum support = 0.05.
For simplicity, let's use
B: Battery
S: Sunscreen
D: sanDals W: boWls
C*1*: {B, S, D, W}
L*1*: {B, S, D, W}
C*2*: {BS, BD, BW, SD, SW, DW}

BW support is 0.025 (< 0.05) so we remove it. SD, SW, and DW supports are unknown, so we remove them
as well.
L2L2: {BS, BD}
C3C3: {BSD}
BSD supports is 0.06 and it's the only itemset remains (no more combinations possible), so it will be the
Frequently Bought Together itemset that we will be used to determine the association rules.
L3L3: {BSD}
Generating Association Rules from Frequent Itemsets
Procedure:
1. For each frequent itemset II in L3L3(in our case, we have one itemset {BSD}), generate all nonempty
subsets of II.
2. For every nonempty subset ss of II, output the rule s→(l−s)s→(l−s), if support(l)support(s)support(l)support(s)
>= minimum confidence threshold,
Now, to proceed we list all the subsets of L3L3:
{B}
{S}
{D}
{BS}
{BD}
{SD}

Confidence(B→SD)=Support(B∧SD)Support(B)=0.060.6=0.1=10%Confidence(B→SD)=Support(B∧SD)Support
(B)=0.060.6=0.1=10%
Confidence(S→BD)=0.060.5=0.12=12%Confidence(S→BD)=0.060.5=0.12=12%
Confidence(D→BS)=0.060.4=0.15=15%Confidence(D→BS)=0.060.4=0.15=15%
Confidence(BD→S)=0.060.1=0.6=60%Confidence(BD→S)=0.060.1=0.6=60%
Confidence(BS→D)=0.060.15=0.4=40%Confidence(BS→D)=0.060.15=0.4=40%
Confidence(B→S)=0.150.6=0.25=25%Confidence(B→S)=0.150.6=0.25=25%
Confidence(B→D)=0.10.6=0.17=17%Confidence(B→D)=0.10.6=0.17=17%
Confidence(S→B)=0.150.5=0.3=30%Confidence(S→B)=0.150.5=0.3=30%
Confidence(D→B)=0.10.4=0.25=25%Confidence(D→B)=0.10.4=0.25=25%
Considering the Minimum Confidence value = 0.25 (25%), we are left with the following interesting rules:
Confidence(BD→S)=0.060.1=0.6=60%Confidence(BD→S)=0.060.1=0.6=60%
Confidence(BS→D)=0.060.15=0.4=40%Confidence(BS→D)=0.060.15=0.4=40%
Confidence(B→S)=0.150.6=0.25=25%Confidence(B→S)=0.150.6=0.25=25%
Confidence(S→B)=0.150.5=0.3=30%Confidence(S→B)=0.150.5=0.3=30%
Confidence(D→B)=0.10.4=0.25=25%Confidence(D→B)=0.10.4=0.25=25%
So the rules, in English, are:
There is 60% chance that a customer will buy Sunscreen, if the customer buy Battery and Sandals
together.
There is a 40% chance that a customer will buy Sandals if the customer buys Battery and Sunscreen
together.
There is a 25% chance that a customer will buy Sandals if the customer buys Battery only.
There is a 30% chance that a customer will buy Battery if the customer buys Sunscreen only.
There is a 25% chance that a customer will buy Battery if the customer buys Sandals only.
We will use Lift rule, to decide on which rule (of the interesting rules) is not coincidental
Lift(X→Y)=Support(X∧Y)Support(X)∗Support(Y)Lift(X→Y)=Support(X∧Y)Support(X)∗Support(Y)
As follows:
Lift(BD→S)=0.060.1∗0.5=1.2Lift(BD→S)=0.060.1∗0.5=1.2
Lift(BS→D)=0.060.15∗0.4=1Lift(BS→D)=0.060.15∗0.4=1
Lift(B→S)=0.150.6∗0.5=0.5Lift(B→S)=0.150.6∗0.5=0.5
Lift(S→B)=0.150.5∗0.6=0.5Lift(S→B)=0.150.5∗0.6=0.5
Lift(D→B)=0.10.4∗0.6=0.42Lift(D→B)=0.10.4∗0.6=0.42
In this case, we could say that the first rule (BD→SBD→S) is the most useful (least coincidental), as it has
highest confidence 60% and Lift 1.2.